src/main/java/soot/FastHierarchy.java

```diff
@@ -346,6 +347,50 @@ public boolean canStoreClass(SootClass child, SootClass parent) {
     }
   }
 
+  /**
+   * "Classic" implementation using the intuitive approach (without using {@link Interval}) to check whether
+   * <code>child</code> can be stored in a type of <code>parent</code>:
+   * 
+   * <p>
+   * If <code>parent</code> is not an interface we simply traverse and check the super-classes of <code>child</code>.
+   * </p>
+   * 
+   * <p>
+   * If <code>parent</code> is an interface we traverse the super-classes of <code>child</code> and check each interface
+   * implemented by this class. Also each interface is checked recursively for super interfaces it implements.
+   * </p>
+   * 
+   * <p>
+   * This implementation can be much faster (compared to the interval based implementation of
+   * {@link #canStoreClass(SootClass, SootClass)} in cases where one interface is implemented in thousands of classes.
+   * </p>
+   * 
+   * @param child
+   * @param parent
+   * @return
+   */
+  protected boolean canStoreClassClassic(final SootClass child, final SootClass parent) {
+    SootClass sc = child;
+    final boolean parentIsInterface = parent.isInterface();
+    while (sc != null) {
+      if (sc.getName().equals(parent.getName())) {
```

Comment:
Why are you comparing the names instead of using object equality on the `SootClasses` directly?
This maybe a bit safer but it's also quite more time consuming to use the String equality here. 
All classes have to be managed by the Scene and are thus guaranteed to be unique. IMHO, if this assumption does not hold, there's an issue elsewhere and should be fixed there. This would  BTW also break many other components of Soot.

Reply:
I took over the class comparison via class name from `RefType.equals()`. Another reason may be hat as an experienced Java developer I encountered dozens of bugs introduced by the use of `==` (on Strings of course which act totally different than a SootClass).

I will change it to an equality and do some more tests.

Reply:
I'm also a huge fan of defensive programming and would, in most cases, also rather tend to use the safer variant of comparing the class names (I guess this is what happens in `RefType.equals()`). However, I think that we are rather going for performance here and should be able to rely on assumptions made across the whole Soot codebase. 
In the end, I can also live with using the safer variant if we the performance penalty compares to the gained safety.

Reply:
I changed the class comparison as you suggested. I did not notice any negative side effects...

Reply:
Thanks!